### PR TITLE
Add a quick check to ensure old version is older than the new version

### PR DESCRIPTION
**Pull request by willdurand**
This will be useful to make sure both #111 and #314 behave correctly. All the other components in `Compare` won't have to deal with the problem of knowing which version is the "old" one and which version is the "new" one. This check happens before any reviewer API calls.

It works because version IDs are integers and newer version IDs get higher numbers (there is some sort of auto-increment). I know we should avoid relying on the auto-increment property of an ID in general, but this is a very simple solution that works well for our purpose. I don't think we'll change the type of version IDs anytime soon (we'll likely always deal with integers) and I don't think we'll change the behavior that currently guarantees that new version IDs have higher numbers than older version IDs. Hence the proposed patch.

The longer version of this (mvp-polish, maybe?) would be to:

- fetch both versions separately (I don't think we need version data in the compare page so it would mean fetching the data only for the comparison)
- compare both versions, but based on what? There is currently no reliable value exposed to perform such a comparison. A `created_at` date could work, if this is something that we store in the database already
- do what's in this patch

If you browse: http://localhost:3000/en-US/compare/502955/versions/1541799...1541798/, this patch detects that the right version is lower than the left version and inverts them + update the URL.

Let me know what you think!

### CODE REVIEW

File: src/pages/Compare/index.tsx [hunk 1]
```diff
@@ -40,13 +41,25 @@ export class CompareBase extends React.Component<Props> {
...
+      );
+      oldVersionId = newVersionId;
+      newVersionId = parseInt(baseVersionId, 10);
+    }
 
     dispatch(
       _fetchVersion({
         addonId: parseInt(addonId, 10),
-        versionId: parseInt(baseVersionId, 10),
+        versionId: oldVersionId,
```
Comment (kumar303) [inline src/pages/Compare/index.tsx]: This is a convenient way to do it but it's confusing because the component is no longer fetching the version requested in the URL. This makes the logic hard to follow and is going to make it hard to debug later on. 

To keep the logic nice and simple, we need to migrate to a `useEffect` hook but until then I would like to see the fetching logic changed to always use the URL parameter. I know it's a bit more complicated and will require more tests but this is more like what we need:

```diff
diff --git a/src/pages/Compare/index.tsx b/src/pages/Compare/index.tsx
index 7f8bc7f..764cca7 100644
--- a/src/pages/Compare/index.tsx
+++ b/src/pages/Compare/index.tsx
@@ -52,16 +52,33 @@ export class CompareBase extends React.Component<Props> {
       history.push(
         `/${lang}/compare/${addonId}/versions/${headVersionId}...${baseVersionId}/`,
       );
-      oldVersionId = newVersionId;
-      newVersionId = parseInt(baseVersionId, 10);
+      return;
     }
 
-    dispatch(
-      _fetchVersion({
-        addonId: parseInt(addonId, 10),
-        versionId: oldVersionId,
-      }),
-    );
+    this.loadData();
+  }
+
+  componentDidUpdate(oldProps: Props) {
+    this.loadData(oldProps);
+  }
+
+  loadData(oldProps?: Props) {
+    const { _fetchVersion, dispatch, match } = this.props;
+    const { addonId, baseVersionId } = match.params;
+    let oldVersionId = parseInt(baseVersionId, 10);
+
+    if (
+      !oldProps ||
+      addonId !== oldProps.match.params.addonId ||
+      baseVersionId !== oldProps.match.params.baseVersionId
+    ) {
+      dispatch(
+        _fetchVersion({
+          addonId: parseInt(addonId, 10),
+          versionId: oldVersionId,
+        }),
+      );
+    }
   }
 
   onSelectFile = () => {};
```

Comment (willdurand) [reply src/pages/Compare/index.tsx]: I wasn't sure because these url params won't be used in the rest of this component and because there is a re-render, the children will receive the versions IDs that match the url (given that the url gets updated directly here) 0

Comment (kumar303) [reply src/pages/Compare/index.tsx]: I'm not sure what you mean. The router prop `props.match.headVersionId` is passed to `_fetchVersion()` but your patch overloads that behavior. Why not always use the router prop and let `history.push()` be the one to change it? That's the patch I proposed.

Comment (willdurand) [reply src/pages/Compare/index.tsx]: Sorry, let me rephrase: I was not sure implementing both `componentDidMount()` and `componentDidUpdate()` was useful because:

- the `baseVersionId` and `headVersionId` are only used in `componentDidMount()`, no need to invert them in `render()` for example (so the confusing behavior is small and scoped)
- as soon as `history.push()` is called, the URL is correct
- the `if` statement inverts the version IDs so that it matches the pushed URL, reflecting the URL seen by the user
- when children are rendered, they will receive the correct values in `baseVersionId` and `headVersionId` (if the children access the router props with `withRouter()`for example)

It's probably a good idea to early return after `history.push()` but this makes me cry already:

```
+    if (
+      !oldProps ||
+      addonId !== oldProps.match.params.addonId ||
+      baseVersionId !== oldProps.match.params.baseVersionId
+    ) {
```

File: src/pages/Compare/index.tsx [hunk 2]
```diff
@@ -38,15 +39,47 @@ export class CompareBase extends React.Component<Props> {
...
+
+  loadData(prevProps?: Props) {
+    const { match } = this.props;
+    const { addonId, baseVersionId, headVersionId } = match.params;
+
+    if (
+      !prevProps ||
+      addonId !== prevProps.match.params.addonId ||
+      baseVersionId !== prevProps.match.params.baseVersionId ||
+      headVersionId !== prevProps.match.params.headVersionId
```
Comment (kumar303) [inline src/pages/Compare/index.tsx]: `headerVersionId` isn't used in the fetch so you don't need to worry about this and you can delete the test.

File: src/pages/Compare/index.tsx [hunk 3]
```diff
@@ -38,15 +39,47 @@ export class CompareBase extends React.Component<Props> {
...
+    ) {
+      const { dispatch, _fetchVersion } = this.props;
+      const oldVersionId = parseInt(baseVersionId, 10);
```
Comment (kumar303) [inline src/pages/Compare/index.tsx]: For the purpose of `fetchVersion()`, I don't think it's necessary to rename this to `oldVersionId`. It doesn't add any helpful context.

File: src/pages/Compare/index.spec.tsx [hunk 4]
```diff
@@ -127,9 +152,192 @@ describe(__filename, () => {
...
+      store,
+    });
+
+    dispatch.mockClear();
+    _fetchVersion.mockClear();
+    root.setProps({
+      match: {
+        params: getRouteParams({
+          addonId,
+          baseVersionId: baseVersion.id,
```
Comment (kumar303) [inline src/pages/Compare/index.spec.tsx]: Neither `baseVersionId` or `headVersionId` seem necessary for this test so if you can move them both to defaults in the helpers then that would make the test easier to read.

File: src/pages/Compare/index.spec.tsx [hunk 5]
```diff
@@ -127,9 +152,192 @@ describe(__filename, () => {
...
+    });
+
+    expect(dispatch).toHaveBeenCalledWith(fakeThunk.thunk);
+    expect(_fetchVersion).toHaveBeenCalledWith({
+      addonId,
+      versionId: baseVersion.id,
+    });
+  });
+
+  it('dispatches fetchVersion() on update if head version is different', () => {
```
Comment (kumar303) [inline src/pages/Compare/index.spec.tsx]: This test looks like it can be deleted since it deals with `headVersionId`.

File: src/pages/Compare/index.spec.tsx [hunk 6]
```diff
@@ -127,9 +152,192 @@ describe(__filename, () => {
...
+      },
+    });
+
+    expect(dispatch).not.toHaveBeenCalled();
+  });
+
+  it('dispatches fetchVersion() on update if base version is different', () => {
+    const addonId = 123456;
+    const baseVersion = fakeVersion;
+    const headVersion = { ...fakeVersion, id: baseVersion.id + 1 };
```
Comment (kumar303) [inline src/pages/Compare/index.spec.tsx]: `headVersionId` doesn't seem relevant to this test so it can be moved to a default in the helpers (if possible).

File: src/pages/Compare/index.spec.tsx [hunk 7]
```diff
@@ -127,9 +152,192 @@ describe(__filename, () => {
     render({
       _fetchVersion,
       store,
+      ...getRouteParams({
+        addonId,
+        baseVersionId: baseVersion.id,
+        headVersionId: headVersion.id,
```
Comment (kumar303) [inline src/pages/Compare/index.spec.tsx]: `headVersionId` doesn't seem relevant to this test

File: src/pages/Compare/index.tsx (hunk 2 above)
Comment (willdurand) [reply src/pages/Compare/index.tsx]: ah sorry I thought I left a comment about it, this will be needed very soon because the compare api requires both ids.

File: src/pages/Compare/index.spec.tsx (hunk 4 above)
Comment (willdurand) [reply src/pages/Compare/index.spec.tsx]: it is actually tricky to remove because the values of `addonId` and `baseVersionId` are used in the `expect()` call right after. `headVersionId` could be omitted though.
